File: vue/src/api/exchange-sources/exchange-sources-manager.js
class RateSourcesManager {
  #rateSources = {};

  registerSource(key, source, caption) {
    if (!key) {
      throw new Error('Source name cannot be null or empty');
    }
    this.#rateSources[key] = { source, caption };
  }

  getRegisteredSources() {
    return Object.getOwnPropertyNames(this.#rateSources).map(
      propertyName => ({ key: propertyName, caption: this.#rateSources[propertyName].caption })
    );
  }

  async getRates(ratesSourceKey) {
    const ratesSource = this.#rateSources[ratesSourceKey].source;
    if (!ratesSource) {
      throw new Error(`Cannot find '${ratesSourceKey}' rates source. Available sources: ${Object.getOwnPropertyNames(this.#rateSources).join(", ")}.`);
    }
    return ratesSource.getRates();
  }
}

const rateSourcesManager = new RateSourcesManager();

export default rateSourcesManager;

File: console_tests/currency-converter-lib/exchange-sources/exchange-sources-manager.cs
namespace CurrencyConverter.ExchangeRateSources {
  public interface IExchangeRatesSource {
    Task<ExchangeRates> GetRates();
  }

  public class ExchangeSourcesManager {
    Dictionary<string, IExchangeRatesSource> sources { get; } = new Dictionary<string, IExchangeRatesSource>();

    public void RegisterSource(string sourceName!!, IExchangeRatesSource exchangeSource!!) {
      if (sourceName == String.Empty) {
        throw new ArgumentNullException($"'{nameof(sourceName)}' cannot be null or empty.", nameof(sourceName));
      }

      this.sources.Add(sourceName, exchangeSource);
    }

    public IExchangeRatesSource GetSource(string sourceName!!) {
      IExchangeRatesSource? exchangeRatesSource;
      if (!sources.TryGetValue(sourceName, out exchangeRatesSource) || (exchangeRatesSource == null)) {
        throw new Exception($"Cannot find '{sourceName}' rates source. Available sources: {String.Join(", ", sources.Keys)}.");
      }
      if (exchangeRatesSource == null) {
        throw new Exception($"A 'null' value is received by the '{sourceName}' rates source name.");
      }
      return exchangeRatesSource;
    }
  }
}
